Add support for HY_GT, HY_LT in query nevra_strict

j-mracek opened the pull request:
It will allow to find upgrades and downgrades directly from nevra.

j-mracek committed:
Add support for HY_GT, HY_LT in query nevra_strict

It will allow to find upgrades and downgrades directly from nevra.

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -1006,11 +1016,17 @@ Query::Impl::filterNevraStrict(int cmpType, const char **matches)
         if (!nevraPattern)
             throw std::runtime_error("Query can not accept NULL for STR match");
-        if (nevraId.parse(pool, nevraPattern)) {
+        if (nevraId.parse(pool, nevraPattern, createNewEVR)) {
```
If `createNewEVR` is `true`, a new string may be created in the pool. Who will remove it from the pool? I think this is not a problem now. Only note.

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -1022,32 +1038,73 @@ Query::Impl::filterNevraStrict(int cmpType, const char **matches)
+                Solvable* s = pool_id2solvable(pool, id);
+                auto low = std::lower_bound(compareSet.begin(), compareSet.end(), *s,
+                                            nameArchCompareLowerSolvable);
+                if (low != compareSet.end() && low->name == s->name && low->arch == s->arch) {
```
It will only work if the `matches` array contains NEVRAs with a unique combination of name and architecture, right? Only the first matching name-arch item from `compareSet` is used.
Probably a filter that contains multiple items with the same name and architecture does not make sense with `HY_LT`/`HY_GT`. And your simplification is not a problem.

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -925,7 +935,7 @@ int
 {
     if (keyname == HY_PKG_NEVRA_STRICT) {
-        if (!(cmp_type & HY_EQ))
+        if (!(cmp_type & HY_EQ || cmp_type & HY_GT || cmp_type & HY_LT))
```
Please fix the method description. There is a statement: "@param cmpType p_cmpType: Allowed compare types - **only HY_EQ or HY_NEQ**"

j-mracek committed:
Do not create new libsolv ID for NEVRA_STRICT filter

In future when daemon will be permanently in memory, creation of a new
id for package query could result in growing usage of memory.

j-mracek committed:
Provide a performance improvement of nevra_strict filter

When only HY_EQ filter is used it provides about 33% performance
improvement.

j-mracek commented:
@rh-atomic-bot test

j-mracek commented:
@rh-atomic-bot try

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -63,20 +63,51 @@ struct default_delete<DnfPackage> {
+NevraID::operator=(const NevraID & src) = default;
+
+NevraID &
+NevraID::operator=(NevraID && src)
```
Why you define your own move assign operator and don't use `default`?

j-mracek replied:
Thanks, removed.

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -63,20 +63,51 @@ struct default_delete<DnfPackage> {
+
+NevraID::NevraID(const NevraID & src) = default;
+
+NevraID::NevraID(NevraID && src)
```
Why you define your own move constructor and don't use `default`?

j-mracek replied:
Thanks, removed.

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -112,8 +130,24 @@ NevraID::parse(Pool * pool, const char * nevraPattern)
+        }
+    } else {
+        size_t evr_size = archDelim - evrDelim;
+        if (evr_size == 0) {
```
`evr_size == 0` is always false, isn't it?
What about to replace 5 lines with one: `evr_str = std::string(evrDelim, archDelim);`

j-mracek commented on libdnf/sack/query.cpp:
```diff
@@ -63,20 +63,42 @@ struct default_delete<DnfPackage> {
+    Id name{0};
+    Id arch{0};
+    Id evr{0};
+    std::string evr_str;
```
Personally I don't like that it initialize empty string and then the object is replaced by another object. The question if I should replace it by char *? It will spare initialization and call evr_str.c_str() when it compares version. @jrohel what do you think?

j-mracek committed:
Strip epoch always when preparing an argument for nevra_strict filter

It will allow to compare two evr and both without 0 epoch.

It uses stripping way haw libsolv does the same job.

j-mracek commented:
@jrohel Fixed, all tests passed

jrohel commented on libdnf/sack/query.cpp:
```diff
@@ -63,20 +63,42 @@ struct default_delete<DnfPackage> {
 namespace libdnf {
 
 struct NevraID {
-    Id name;
```
Code can be inline. We need initialize Ids to "0" only in default (non parameterized) constructor.
```
    NevraID() : name(0), arch(0), evr(0) {};
    NevraID(const NevraID & src) = default;
    NevraID(NevraID && src) noexcept = default;
    NevraID & operator=(const NevraID & src) = default;
    NevraID & operator=(NevraID && src) = default;
    Id name;
    Id arch;
    Id evr;
    std::string evr_str;
```

jrohel commented:
Super. Nice replacement of the slow Python DNF code. I wrote one note to `NevraID`. Methods can be defined inline. You can define they directly in class declaration. Initialization of `Id`s is needed only in default (nonparameterized) constructor. Move constructor can be marked `noexcept` for DNF4.
In DNF5 (C++17) we will mark more methods `noexcept` and some of them also `constexpr`.
